### ISSUE TITLE
AADHttpClient authentication fails in MS Teams desktop client after user changed password

### ISSUE THREAD

**Issue by NickSevens** [SharePoint/sp-dev-docs#5320]
## Category

- [x] Bug

## Describe the bug

We are loading an SPFx webpart into a personal MS Teams app. When using this application in the **desktop client**, when using AADHttpClient in the webpart, the auth flow fails after a user has changed his password, even after waiting a couple of days.
In the web client the issue doesn't occur.

The issue occurs in the call to `Microsoft.SharePoint.Internal.ClientSideComponent.Token.AcquireOBOToken`, which returns the following error:

> {"odata.error":{"code":"10001","message":{"lang":"en-US","value":"AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2019-12-18T13:53:05.0728393+00:00' and the TokensValidFrom date (before which tokens are not valid) for this user is '2020-02-06T08:34:15.0000000Z'.\r\nTrace ID: f5a28c6d-bd35-4f3c-a0e5-e65d5d261900\r\nCorrelation ID: 855d349f-405f-b000-e134-72e59f622787\r\nTimestamp: 2020-02-11 09:54:02Z"},"error.redirectUrl":"https://<tenant>.sharepoint.com/...?reauthid=2.MXwwfA.Ec3DKPGt_1Gm0Ag....."}}

**Comment by Ollzi** [comment]
We have a customer where one of the users is experiencing the exact same issue.

**Comment by andrewconnell** [comment]
Did the user logout/login after changing their password? Seems the access token is cached from the initial login & changing the password invalidated the token.

**Comment by NickSevens** [comment]
I tried it all :( logout/login, quit Teams, reboot, change to developer mode and back, etc ...
No luck ...

What helped in the end was explicitly navigating to the URL provided in `error.redirectUrl` within teams (through the debug console). This is however not something we can let all end-users do whenever they have the issue...

Seems like this could and should be handled internally when it occurs.

**Comment by andrewconnell** [comment]
eek... agreed, that's not a great experience...

/cc @patmill @lucabandMSFT - is this a known thing or something you were looking into with the improvements to SPFx & the MSTeams integration?

**Comment by Ollzi** [comment]
@NickSevens, our customer is authenticating via ADFS, is it the same setup in your case?

**Comment by NickSevens** [comment]
@Ollzi we used to but not anymore.

**Comment by Ollzi** [comment]
@patmill, @lucabandMSFT, @andrewconnell we have investigated this further with our customer and it was solved with the solution you suggested @NickSevens.

**The following error information is returned in the error message when this happens:**

{odata.error: {code: "10001", message: {lang: "sv-SE",…},…}}
odata.error:{code: "10001", message: {lang: "sv-SE",…},…}
code:"10001"
error.redirectUrl:"https://xxxxx.sharepoint.com/sites/provisionering/SitePages/Provisioning.aspx?reauthid=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
message: {lang: "sv-SE",…}
lang:"sv-SE"
value:"AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2020-01-08T08:31:15.0750000Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2020-01-16T07:57:14.0000000Z'."

**Comment by lucabandMSFT** [comment]
adding @lahuey  for awareness... Glad that was solved. indeed that's not something customers can do to solve the issue..

**Comment by NickSevens** [comment]
@lucabandMSFT @andrewconnell thanks for the followup.

We've indeed managed to help most people with the workaround mentioned above, but I do hope there will be a fix implemented soon.
Company policy makes us change passwords pretty often, so we can't expect users to do this every 2-3 months :)

@ollzi glad it helped for you as well.

**Comment by andrewconnell** [comment]
@NickSevens said:
> we can't expect users to do this every 2-3 months

If the link is provided in the error object, wouldn't you then be able to detect it & handle it in the solution?

**Comment by NickSevens** [comment]
@andrewconnell the link is provided in the error response of `https://<tenant>.sharepoint.com/_api/Microsoft.SharePoint.Internal.ClientSideComponent.Token.AcquireOBOToken?resource=<guid>&clientId=<guid>`, which is not something we call ourselves, but I assume is coming from within `AadHttpClient` - and is only called on Teams desktop client (and if I recall correctly also in Android/iPhone apps)
Seems to me this should be handled internally by SPFx, to acquire a correct token?

**Comment by andrewconnell** [comment]
Agreed... seeing `Internal` in the endpoint says _"not for use by 3rd party... it's MSFT so..."_

![](https://media.giphy.com/media/hxc32veg6tbqg/giphy.gif)

**Comment by Ollzi** [comment]
Any updates on this?

**Comment by NickSevens** [comment]
@andrewconnell do you have any update on this?
We saw the same issue occurring with an MFA being expired. The "solution" redirect URL was again in the response message, so using that message to redirect the user again fixed the issue.
This again feels like something that could be done automatically by the calling function when such a response is received.

**Comment by andrewconnell** [comment]
Nope... I'm not in MSFT engineering so it's not for me to share updates. @lucabandMSFT & @lahuey are on the thread... any updates would be more appropriate to come from them.

**Comment by NickSevens** [comment]
@lucabandMSFT @lahuey any update on this in the meantime?

**Comment by NickSevens** [comment]
@lucabandMSFT @lahuey We're encountering this issue at multiple customers now as well. Explaining the end-users to open up dev tools and so on to fix it, is really annoying.
Is there any progress/movement here? Something we can tell our customers in the meantime?

**Comment by joebergqvist** [comment]
I have several customers with conditional access setup with login session time limited. I can easily reproduce this issue using a very narrow window of login time. 

The refreshtoken is cached and should be replaced when logging out and in in the client itself. This is a really bad experience. 

This should be handled internally as several others stated. 
If not how can you script a redirect against this url without hitting cors issues? Ive been trying with crossdomainrequest method. The only thing that seems to be working but is terrible is opening the redirect as a deeplink which opens a browser.

Kind regards johan

**Comment by PieterVandendriessche** [comment]
Is there any update on this issue? Fix seems not to work in desktop client of MacOS.

**Comment by lucabandMSFT** [comment]
Folks, SUPER SORRY for the delay in answering here. I've an update.. it's _some news__....:
That's unfortunately by design for now:
**Reason**: when acquiring Tokens to access external resources from Teams Desktop Client, we use a flow grant called "onBehalfOf", which basically relies on exchanging a token that was issued for one resource to acquire the token to access another resource.

In this case, the first resource is SharePoint and we generate that token when the user logs-in to SharePoint.

Upon password change / reset, the existing token that was issued to SharePoint is no longer valid and the user needs to re-login to SharePoint to update the token used to complete the OBO token mentioned above.

That is the same reason why, today, if you use a web part on Teams rich client that needs tokens to access a particular resource (e.g. Microsoft Graph) but you never accessed SharePoint first, that request will fail.

We are working very hard with the AAD identity team to move to MSAL V2 identity library and adopt the "Code Authorization" grant flow. Once that is in place, there will be no longer reasons to use the SharePoint token to acquire tokens to access different resources.

We unfortunately do not have a solid ETA yet. But we understand the scenario is real, common, and quite frustrating right now: that's why we are very much committed to address it ASAP.

**Comment by petecuttriss** [comment]
Hi @joebergqvist, We are battling with this issue also.  We have customers who are using a Teams first approach and as such aren't logging into SharePoint prior to using Teams.  The SPFx web part that we have in Teams that calls the graph fails as the user hasn't logged into SharePoint previously.

Is there any chance you could share the workaround code that you are using to get the redirect url and then opening it via the executeDeeplink function?

@lucabandMSFT, I appreciate all the efforts going in to sort out this issue.

Kind regards
Pete

**Comment by cloudhal** [comment]
We are having this same issue, also developed a Teams first approach where we don't want users going to sharepoint, everything is done through Teams. We have a sharepoint site with a custom web part, it does not work after a user has changed their password and they have to go to the site and login, then it will work until the next password change. This means that using Teams for everything is not really a viable strategy unfortunately...

**Comment by NickSevens** [comment]
@lucabandMSFT we now see the same behavior also happening **in the browser**, whereas previously it only happened in the fat clients.
It seems something has changed, and now **AcquireOBOToken** is also called in the browser, which then fails after a user has changed their password. This change is visibile on multiple customer tenants already.
Navigating to SharePoint unfortunately doesn't work as a workaround.
The error message in the request however shows a different message than before:
**"Missing refresh token"**
```
{
  "odata.error": {
    "code": "10001",
    "message": {
      "lang": "fr-FR",
      "value": "Missing refresh token."
    },
    "error.redirectUrl": "https://<tenant>.sharepoint.com/sites/<sitecollection>/_layouts/15/teamshostedapp.aspx?teams&personal&componentId=ea7bc92f-4215-4778-80ec-ebc37f52b378&forceLocale=fr-fr&env=TeamsWebView&reauthid=2.MXwwfA.X_j4N5HfjtlLsL1tBTbKfG3Dk4UBT..."
  }
}
```

**Comment by GrahamMcMynn** [comment]
@NickSevens - thanks for reporting that issue.  We had a regression which caused the client to use AcquireOBOToken in the browser and it is now being fixed.

**Comment by NickSevens** [comment]
@GrahamMcMynn glad it's confirmed :) any timeframe on the fix we can communicate to our customers?

**Comment by GrahamMcMynn** [comment]
@NickSevens - it will still be about 3 weeks minimum before a fix is rolled out to production as it did not make it for last weeks build.

**Comment by NickSevens** [comment]
@GrahamMcMynn it seems the _browser_ calls don't happen anymore. Could you confirm it's been rolled out to all tenants?

**Comment by doxic012** [comment]
The issue still occurs to one of my customers when the MFA token expires or the user is forced to log out via admin panel.

**Comment by GrahamMcMynn** [comment]
We have not fixed the desktop client issue, so it is expected to still be happening.  We are looking into solutions for that still.  The fix that rolled out was for the regression where this was happening in the browser.

**Comment by WardWilmsen** [comment]
1 of our customers is experiencing the same issue, but it will even break without a password change... users will be able to use the Teams app after logging into SharePoint. But this works for a few days (2 - 3) and the issue will come back.

This has a heavy impact on our customer, they paid for the Teams app, but can not use it. For 6 other customers it's already working fine for the past few months... so not sure what is going on there. 

Opened a MS Premier support ticket on this, they told me to post here and wait for a solution from the MS dev team. 😕

**Comment by doxic012** [comment]
@WardWilmsen maybe this helps you out:
I did a workaround in our (SPFx-) app in the init/render parts by loading the current user context (`api.("me").get()`).
The link will be opened directly, but I'm also displaying it in a MessageBar above the actual component
(https://developer.microsoft.com/en-us/fluentui#/controls/web/messagebar)

```typescript
protected asnyc onInit() {
  try {
    this.graphClient = await this.context.msGraphClientFactory.getClient();
    this.user = await this.graphClient.api("me").get();
  } catch (err) {
    const errorPromise = await err; // transform promise into actual error
    console.error("Error on loading graph client", errorPromise);

    if (errorPromise["odata.error"]) {
      const odataError = errorPromise["odata.error"];
      const redirectUrl = odataError["error.redirectUrl"];

      if (redirectUrl) {
        // This opens a new browser window for the user to follow the link
        // From my current experience it seems to work but takes some time to take effect,
        // so I stored the redirectUrl and displayed it above the acutal component in a MessageBar
        this.context.sdks.microsoftTeams.teamsJs.executeDeepLink(redirectUrl);
        this.errorUrl = redirectUrl;
      }
    }
  }
}

public render(): void {
  const errorElement = this.errorUrl && React.createElement(ErrorView, {errorUrl: this.errorUrl});
  const appElement = React.createElement(App, {
    ...
  });
  ReactDOM.render(React.createElement(React.Fragment, null,
    errorElement,
    appElement
  ), this.domElement);
}
```

**Comment by WardWilmsen** [comment]
@doxic012 thx a lot for your suggestion, we'll give this a try in a few days. I'll post an update here once I have the result.

**Comment by WardWilmsen** [comment]
@doxic012 we tried this, but this does open a new browser window with the default browser right? I thought it would be able to redirect within the Teams tab, but that won't work as there is an iframe blocker for the login.microsoftonline.com domain.

Refused to display https://login.microsoftonline.com/...' in a frame because it set 'X-Frame-Options' to 'deny' if we try to redirect within the teams tab. Any suggestion?
Because opening a new browser window for this is causing way too much distraction & confusion for the end users in my opinion.

**Comment by doxic012** [comment]
@WardWilmsen I found no other way than doing this, as location.href, etc. always forbid opening the link due to the X-Frame-Options.
so `this.context.sdks.microsoftTeams.teamsJs.executeDeepLink` was the only way to open the link from the app itselves.

I also added a MessageBar from fluentui/react with an info-message and a Link-Element directing to the redirectUrl.

```typescript
<MessageBar messageBarType={MessageBarType.warning} isMultiline>
      An error occurred while loading context data for the Microsoft Graph API.
      This is a known problem due to an invalid token from multi factor authentication. Click here to <Link href={props.errorUrl} target={"_blank"}>generate a new access token</Link> (This might take a while to take effect).
</MessageBar>
```
Maybe that's an option for you?

**Comment by NickSevens** [comment]
@lucabandMSFT we're seeing this issue with a custom in the desktop client again, now showing "Missing refresh token" as error message. This seems a _bit_ different than previously, where the issue was usually correlated to a changing password or MFA setup.
There's no such relation in this instance however.

Is there anything we can do? The usual workarounds (opening SharePoint & logout/login ; navigating to the `error.redirectUrl`) don't seem to work for this case.

**Comment by Rowdy** [comment]
@lucabandMSFT any updates on this behavior. We'd love to roll out the first custom made personal Teams apps company wide, but as long as this is a known bug we cannot proceed any further.

**Comment by Rowdy** [comment]
> EDIT:
> Maybe a forced HTTP Get call to the URL might work? I'm not sure if it will trigger the update on the expired AAD token.
Has anyone tried that already?

**Comment by NickSevens** [comment]
@Rowdy as far as our investigations go, the `error.redirectUrl` needs to be visited in a browser, to make authentication work. Redirecting the user to this URL in the Teams desktop client frame doesn't seem to work, unfortunately.
I've also tried opening the URL in the `microsoftTeams.authentication.authenticate` popup frame, but no token seems to be cached for the user at that point. Quite strange behavior, since SP **is** loading in the popup frame - so no sure what the difference with loading in a browser is...

**Comment by clechner77** [comment]
Visiting the link in a browser doesn't work now.  A vanilla spfx web part doing a basic graph call fails on the Teams desktop client.  Doesn't matter if a user has changed password or not, this is all users.  Granted most in my case are MFA, which apparently didn't work either.  Any workarounds or is it just dead in the water now?

**Comment by lucabandMSFT** [comment]
Hello everyone
I want to provide an update on this topic: I know that we have several issues relate to SPFx components in Teams Rich client, and we have mentioned before we were working on that: time to provide some updates.

Couple of things happened in the past months:

- First, we got rid of ADAL library client side and we replaced it with MSAL V1. Now all SPFx solutions (existing, new, 1st party and 3rd party) are using MSAL library when performing implicit grant to acquire a token from AAD using AADHTTPClient library or MSGraphClient library
- Second, we introduced logic to detect and work with Browser cookie blocking technologies (like Safari ITP) while waiting for MSAL V2 implementation (which is also something we are working right now); 
- Third, we have implemented a new OnBehalfOf (OBO) flow that we will use when we detect that the SharePoint Framework component runs on Teams Desktop or Teams mobile. The reference implementation is something similar to what Teams describes here (https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/auth-aad-sso) with some additional work needed to integrate with the admin consent model and work needed for 1st, 3rd and isolated components. But the point is that such model will get rid of any "you need to authenticate to SharePoint first" requirements.

Before we can start enabling this model across all the tenants a couple of things needs to be done:

1. We need to finish coding: we are at a point where the end to end flow works and it is stable. We are now working to enable conditional access policy support and cross tenant accounts
2. Testing: everything around auth is hard, complex and extremely powerful so we need to be sure we don't leave have any entry point for potential token spoofing attack. That's why it takes such long time for us to ship anything new or fix anything in regards of auth.

So that's pretty much it at this point. We are working on this and while we cannot commit to a precise date, I can tell you that we are actively working on it and I hope (and expect) to be able to provide some additional updates soon.

**Comment by NickSevens** [comment]
Thank you for the update @lucabandMSFT. It's nice to hear some progress for this. :)

**Comment by vysta** [comment]
@lucabandMSFT & @VesaJuvonen has there been any additional progress or update regarding timing?

**Comment by lucabandMSFT** [comment]
the fix is rolling.. unfortunately it takes time to the Teams client to update and we are waiting for the bits to land on Teams before turning the capability on in SharePoint.

I unfortunately don't have a solid timeline to share with you but we will provide an update as soon as the bits are available in Teams client.

**Comment by NickSevens** [comment]
That's great news @lucabandMSFT ! Looking forward to it :) We've all been waiting for a - totally understandable - long time :)

**Comment by VesaJuvonen** [comment]
Quick update on this - Fix should be enabled in production for Teams and SharePoint within upcoming days

**Comment by clechner77** [comment]
@VesaJuvonen does this include MFA as wel?  I would assume so...

**Comment by NickSevens** [comment]
@VesaJuvonen @lucabandMSFT we seem to be getting a different kind of error now in MS Teams, which I think is related to the upgrade as well.
I can see there have been changes to the _chunk.adalauthcontext....js_ file, which now loads to `window.webpackJsonpc07208f0_ea3b_4c1a_9965_ac1b825211a6_1_13_0` instead of `...1_12_1`.

The ` AcquireOBOToken` flow now requests a token with a different `ClientId` (`08e18876-6177-487e-b8b5-cf950c1e598c`), but when loading a SharePoint page (in a `SharePoint` Teams tab), the promise to fetch a token is never resolved.

https://github.com/SharePoint/sp-dev-docs/issues/7248 states the rollout should be completed, but something still seems off :/

Some more info also at https://github.com/SharePoint/sp-dev-docs/issues/7307

**Comment by clechner77** [comment]
> @VesaJuvonen @lucabandMSFT we seem to be getting a different kind of error now in MS Teams, which I think is related to the upgrade as well.
> I can see there have been changes to the _chunk.adalauthcontext....js_ file, which now loads to `window.webpackJsonpc07208f0_ea3b_4c1a_9965_ac1b825211a6_1_13_0` instead of `...1_12_1`.
> 
> The ` AcquireOBOToken` flow now requests a token with a different `ClientId` (`08e18876-6177-487e-b8b5-cf950c1e598c`), but when loading a SharePoint page (in a `SharePoint` Teams tab), the promise to fetch a token is never resolved.
> 
> #7248 states the rollout should be completed, but something still seems off :/
> 
> Some more info also at #7307

@NickSevens - try visiting the Sharepoint API Access page...  https://XXXXXXX-admin.sharepoint.com/_layouts/15/online/AdminHome.aspx#/webApiPermissionManagement - might take a bit, then see if the desktop client magically works (not sure if a close/re-open is necessary)

**Comment by NickSevens** [comment]
Well, it depends:
* In a `SharePoint` teams tab, I get **nothing** (the call to AcquireOBOToken is never fired)
* In a `Website` teams tab, pointing to the SharePoint URL, I get a 500
* In a `Personal App`, pointing to the `TeamsLogon.aspx?dest=...` page, with SSO setup, I get a 500 as well

Navigating to the API permissions page didn't change anything.

_PS: in the browser, everything still works fine_

**Comment by lucabandMSFT** [comment]
@NickSevens, two questions:

- When you navigate to the APi Access page in SPO tenant admin, can you use an user that has both SPO Tenant Admin role and AAD Global Admin role?
- does your option 1 in the list above (SharePoint Teams tab) work? or it also fails?

**Comment by GrahamMcMynn** [comment]
@NickSevens - Sorry one more question for you.  How was the teams tab created for the page?  Was it from adding a page via the first party app that already existed, or did you use a custom teams tab that you made yourself?  If it was from the latter experience (you created the manifest.json yourself) you will need to ensure you are using the {teamSiteDomain} moniker in order for the token to resolve.  That needs to be used under webapplicationinfo -> resource.  For example:
  "webApplicationInfo": {
    "resource": "https://{teamSiteDomain}",
    "id": "00000003-0000-0ff1-ce00-000000000000"
  }

**Comment by doxic012** [comment]
Bump for this issue, so it doesn't close.

The issue still occurs sporadically in an spfx webpart using the following supportedHosts-config, being deployed to teams via the the apps for sharepoint-list:
`"supportedHosts": ["TeamsPersonalApp"]`

I am using both of the following methods to instantiate a graph-client to get some user data, etc.:

```typescript

   // Using this.context
   this.graphClient = await this.context.msGraphClientFactory.getClient();
   const me = await this.graphClient.api(`me`).version("v1.0").get();

   // Using serviceScope:
   this.scope.createDefaultAndProvide(MSGraphClientFactory.serviceKey);
   ...
   const graphClientFactory = scope.consume(MSGraphClientFactory.serviceKey);
   const graphClient = await this.graphClientFactory.getClient();
   await graph.api(...)

```

**Comment by lucabandMSFT** [comment]
@doxic012, is the manifest part of the solution? or the manifest in Teams is created by using "Sync to Teams" option in SharePoint App Catalog?

**Comment by doxic012** [comment]
> @doxic012, is the manifest part of the solution? or the manifest in Teams is created by using "Sync to Teams" option in SharePoint App Catalog?

@lucabandMSFT 
It's automatically generated using "Sync to Teams".
I also tried to do it with a custom manifest and installed it from the teams admin center, but that didn't change anything, so I went back to the sync option

**Comment by lucabandMSFT** [comment]
@doxic012, thanks. Let's please open an new issue to track the current problem you are experiencing as we want to be sure we are not confusing with other issues.
The initial issue that was reported here has been fixed long time ago and the other issue that @NickSevens raised on September 14th, has been fixed too.

This seems a completely new one which is why I'm asking you to please open an new issue and provide details about the situation (e.g. is that happening in desktop? web? mobile?).

Thanks a lot!
